firebase_push_notifications/firebase_push_notifications/v1_0/manager.py

```diff
@@ -0,0 +1,148 @@
+"""This module contains the logic for sending push notifications to the user's device using the Firebase Cloud Messaging service."""  # noqa: E501
+import json
+import logging
+import os
+from datetime import timedelta
+
+import google.auth.transport.requests
+import requests
+from aries_cloudagent.messaging.util import datetime_now, time_now
+from aries_cloudagent.storage.base import StorageNotFoundError
+from dateutil import parser
+from google.oauth2 import service_account
+
+from .constants import (
+    BASE_URL,
+    ENDPOINT_PREFIX,
+    ENDPOINT_SUFFIX,
+    MAX_SEND_RATE_MINUTES,
+    SCOPES,
+)
+from .models import FirebaseConnectionRecord
+
+PROJECT_ID = os.environ.get("FIREBASE_PROJECT_ID", "")
+FCM_ENDPOINT = ENDPOINT_PREFIX + PROJECT_ID + ENDPOINT_SUFFIX
+FCM_URL = BASE_URL + "/" + FCM_ENDPOINT
+
+LOGGER = logging.getLogger(__name__)
+
+
+def _get_access_token():
+    """Gets an access token for sending firebase messages with the service account credentials."""  # noqa: E501
+    credentials = service_account.Credentials.from_service_account_info(
+        json.loads(os.environ.get("FIREBASE_SERVICE_ACCOUNT")), scopes=SCOPES
+    )
+    request = google.auth.transport.requests.Request()
+    credentials.refresh(request)
+    return credentials.token
+
+
+async def send_message(profile, connection_id):
+    """Attempt to send a push notification to the device associated with the connection id if passes the following checks."""  # noqa: E501
+    headers = {
+        "Authorization": "Bearer " + _get_access_token(),
+        "Content-Type": "application/json; UTF-8",
+    }
+
+    async with profile.session() as session:
+        record = None
+        try:
+            record = await FirebaseConnectionRecord.retrieve_by_connection_id(
+                session, connection_id
+            )
+        except StorageNotFoundError as e:
+            LOGGER.debug(
+                f"Could not retrieve device token for connection {connection_id} : {e}"
+            )
+            return
+
+        """ Don't send token if it is blank. This is the same as being disabled """
+        if not record or record.device_token == "":
+            return
+
+        """ 
+            To avoid spamming the user with push notifications, 
+            we will only send a push notification if the last one 
+            was sent more than MAX_SEND_RATE_MINUTES minutes ago. 
+        """
+        if record.sent_time is not None and parser.parse(
+            record.sent_time
+        ) > datetime_now() - timedelta(minutes=MAX_SEND_RATE_MINUTES):
+            LOGGER.info(
```

Comment:
Depending on how "chatty" we want the mediator service to be, statements like this might be better as `debug` messages.